src/index.tsx

```diff
@@ -0,0 +1,140 @@
+import invariant from 'invariant';
+import React, { useCallback, useContext, useState } from 'react';
+import { ReactRelayContext, commitMutation } from 'react-relay';
+import {
+  MutationConfig as BaseMutationConfig,
+  Environment,
+  OperationBase,
+} from 'relay-runtime';
+
+import { Omit, WithOptionalFields } from './typeHelpers';
+
+export type MutationState<T extends OperationBase> = {
+  loading: boolean;
+  data: T['response'] | null;
+  error?: Error | null;
+};
+
+type MutationConfig<T extends OperationBase> = WithOptionalFields<
+  Omit<BaseMutationConfig<T>, 'mutation'>,
+  'variables'
+>;
+
+export type Mutate<T extends OperationBase> = (
+  config?: Partial<MutationConfig<T>>,
+) => void;
+
+export function useMutation<T extends OperationBase>(
+  mutation: BaseMutationConfig<T>['mutation'],
+  userConfig: MutationConfig<T> = {},
+  environment?: Environment,
+): [Mutate<T>, MutationState<T>] {
+  const [state, setState] = useState<MutationState<T>>({
+    loading: false,
+    data: null,
+    error: null,
+  });
+
+  const relayContext = useContext(ReactRelayContext);
+  const resolvedEnvironment = environment || relayContext.environment;
+  const {
+    configs,
+    variables,
+    uploadables,
+    onCompleted,
+    onError,
+    optimisticUpdater,
+    optimisticResponse,
+    updater,
+  } = userConfig;
+
+  const mutate = useCallback(
+    (config?: Partial<MutationConfig<T>>) => {
+      const mergedConfig = {
+        configs,
+        variables,
+        uploadables,
+        onCompleted,
+        onError,
+        optimisticUpdater,
+        optimisticResponse,
+        updater,
+        ...config,
+      };
+
+      invariant(mergedConfig.variables, 'you must specify variables');
```

Comment:
i think technically mutations without variables are valid. though not very useful

Reply:
the upstream type requires variables